pkg/executor/build.go

```diff
@@ -741,6 +734,94 @@ func CalculateDependencies(stages []config.KanikoStage, opts *config.KanikoOptio
 	return depGraph, nil
 }
 
+func MakeStageBuilders(stages []config.KanikoStage, opts *config.KanikoOptions, args *dockerfile.BuildArgs, crossStageDeps map[int][]string, dcm map[string]string, stageNameToIdx map[string]int, fileContext util.FileContext) ([]*stageBuilder, error) {
+	images := make(map[int]v1.Image)
+	stageIdxToCacheKey := make(map[int]string)
+
+	var stageBuilders []*stageBuilder
+
+	for _, s := range stages {
+		ba := dockerfile.NewBuildArgs(opts.BuildArgs)
+		ba.AddMetaArgs(s.MetaArgs)
+		var image v1.Image
+		var err error
+		if s.BaseImageStoredLocally {
+			image = images[s.BaseImageIndex]
+		} else if s.Name == constants.NoBaseImage {
+			image = empty.Image
+		} else {
+			image, err = image_util.RetrieveSourceImage(s, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		sb, err := newStageBuilder(
+			image,
+			args, opts, s,
+			crossStageDeps,
+			dcm,
+			stageIdxToCacheKey,
+			stageNameToIdx,
+			fileContext)
+		if err != nil {
+			return nil, err
+		}
+
+		args = sb.args
+		// Set the initial cache key to be the base image digest, the build args and the SrcContext.
+		var compositeKey *CompositeCache
+		if cacheKey, ok := sb.digestToCacheKey[sb.baseImageDigest]; ok {
+			compositeKey = NewCompositeCache(cacheKey)
+		} else {
+			compositeKey = NewCompositeCache(sb.baseImageDigest)
+		}
+
+		// Apply optimizations to the instructions.
+		if err := sb.optimize(*compositeKey, sb.cf.Config); err != nil {
+			return nil, errors.Wrap(err, "failed to optimize instructions")
+		}
+		images[s.Index] = image
+		stageBuilders = append(stageBuilders, sb)
+		stageIdxToCacheKey[s.Index] = sb.finalCacheKey
+	}
+	return stageBuilders, nil
+}
+
+func filterOnBuild(cmds []commands.DockerCommand) []commands.DockerCommand {
+	var out []commands.DockerCommand
+	for _, c := range cmds {
+		switch cmd := c.(type) {
+		case *commands.OnBuildCommand:
+			// Skip ONBUILD commands
+		default:
+			out = append(out, cmd)
+		}
+	}
+	return out
+}
+
+func squash(a, b *stageBuilder) *stageBuilder {
+	acmds := filterOnBuild(a.cmds)
+	return &stageBuilder{
```

Comment:
we should split stageBuilder into two parts. It currently has two responsibilities, it cares both `how` to build and `what` to build. I think the `how` can stay in stageBuilder, but the `what` should be a separate struct entirely. This new struct can then be passed into the `build()` function. Then the same stagebuilder entity can crunch through all stages. Here in the squash it becomes obvious as only the `what` parts are affected by the squashing, for the `how` parts squashing is meaningless.